solver/src/main/java/org/chocosolver/solver/search/strategy/selectors/variables/ConflictHistorySearch.java

```diff
@@ -0,0 +1,133 @@
+
+/**
+ * Source: "Conflist History Based Branching Heuristic for CSP Solving", Habet and Terrioux.
+ * <p>
+ * Project: choco.
+ *
+ * @author Charles Prud'homme
+ * @since 25/02/2020.
+ */
+@SuppressWarnings("rawtypes")
+public class ConflictHistorySearch
+    extends AbsctractCriterionBasedStrategy
+    implements IMonitorContradiction, IMonitorRestart {
+
+    /**
+     * Decreasing step for {@link #a}.
+     */
+    private static final double STEP = 10e-6;
+    private static final double D = 10e-4;
+    private static final double DECAY = .995;
+    /**
+     * Score of each propagator.
+     */
+    private final TObjectDoubleMap<Propagator> q = new TObjectDoubleHashMap<>(10, 0.5f, 0.0);
+    /**
+     * Step-size, 0 < a < 1.
+     */
+    private double a = .4d;
+    /**
+     * Reward value of each propagator.
+     */
+    private final TObjectDoubleMap<Propagator> r = new TObjectDoubleHashMap<>(10, 0.5f, 0.0);
+    /**
+     * The number of conflicts which have occurred since the beginning of the search.
+     */
+    private int conflicts = 0;
+    /**
+     * Last {@link #conflicts} value where a propagator led to a failure.
+     */
+    private final TObjectIntMap<Propagator> conflict = new TObjectIntHashMap<>(10, 0.5f, 0);
+
+
+    public ConflictHistorySearch(IntVar[] vars, long seed, IntValueSelector valueSelector) {
+        super(vars, seed, valueSelector);
+        vars[0].getModel().getSolver().plugMonitor(this);
+    }
+
+    @Override
+    public boolean init() {
+        Solver solver = vars[0].getModel().getSolver();
+        if(!solver.getSearchMonitors().contains(this)) {
+            vars[0].getModel().getSolver().plugMonitor(this);
+        }
+        return true;
+    }
+
+    @Override
+    public void remove() {
+        Solver solver = vars[0].getModel().getSolver();
+        if(solver.getSearchMonitors().contains(this)) {
+            vars[0].getModel().getSolver().unplugMonitor(this);
+        }
+    }
+
+
+    @Override
+    public void onContradiction(ContradictionException cex) {
+        if (cex.c instanceof Propagator) {
+            Propagator p = (Propagator) cex.c;
+            double qj = q.get(p);
+            // update q
+            q.put(p, (1 - a) * qj + a * r.get(p));
+            // update r
+            double rj = r.get(p);
+            r.put(p, 1d / (conflicts - rj + 1));
+            // decrease a
+            a = Math.max(0.06, a - STEP);
+            // update conflicts
+            conflict.put(p, conflicts);
+            conflicts++;
+        }
+    }
+
+    @Override
+    protected double weight(IntVar v) {
+        double w = 0.;
+        int nbp = v.getNbProps();
+        for (int i = 0; i < nbp; i++) {
+            Propagator prop = v.getPropagator(i);
+            int pid = prop.getId();
+            int futVars = pid2arity.get(pid);
+            if (futVars == -1) {
+                futVars = prop.arity();
+                pid2arity.put(pid, futVars);
+            }
```

Comment:
I do not get the idea behind pid2arity. In the paper, the weight is increased by q.get(prop) + D only if its number of uninstantiated variables is strictly greater than one. Shouldn't we do the same behaviour with a method that would look like this ?

`
private static boolean hasStrictlyMoreThanOneUninstantiatedVariables(Propagator prop) {
        int nbUninstantiatedVariables = 0;
        for(int i = 0; i<prop.getNbVars(); i++) {
            if(!prop.getVar(i).isInstantiated()) {
                nbUninstantiatedVariables++;
                if(nbUninstantiatedVariables > 1) {
                    return true;
                }
            }
        }
        return false;
    }
`

Or is it too costful on the long run ?

Reply:
There is something to do here, I agree. I will suggest something else.